Bugs and 1.13+

Jakllp opened the pull request:
Fixed:
- Compiling 1.16 through maven
- Rail-Detection (falling of rail-edge - rollercoasters are rolling again)
- Config now supports "black_wool" instead of "WOOL:15" as road-blocks
- Some Legacy-Stuff
- Dropping Cart+Content (e.g. Storage-Cart)

storm345 reviewed (changes requested):
Pretty much same feedback as on the preview pull request:
https://github.com/storm345dev/ucars/pull/8

Notably:
-Compatibility with 1.12
-Not all uCars are Minecarts, you assumed they were
-Ignoring custom car hitboxes, would break block climbing for cars with custom hitboxes defined via uCarsTrade
-Existing configs need to still work or be converted to different format.
-Not sure why you change the road block check to pass the material name uppercased instead of the block, would be preferable to pass the block so that you are not breaking the abstraction of 'block' and 'config identifier for block' unnecessarily (fragile code)
-Removing air from being considered a road block likely breaks vehicles which are falling or similar? Is this not an issue?

storm345 reviewed (commented):
Your 'getBehindBlock()' is also a weird implementation that seems to rely on velocity being lower than a threshold and doesn't look at first glance like it would work well for diagonals.

Jakllp commented:
Basically what I expected: Things will break when in use with other plugins like uCarsTrade.

Some things I want to note: 
The „getBehindBlock()“ seems to work fine (haven‘t checked on diagonal rail lines so good point), gonna do some more testing
The air as a road block broke some other things that were fixed with it not being considered but will double check.
I seemingly need to pass the roadblock that way as... when I pass the block and use the same method to get the name of the block it gives WOOL 15 and not black_wool
Funnily enough your placing function expects black_wool as a road block instead of wool:15.

As for the custom hitboxes... I have no clue why and when I deleted that

Thanks for the heads up with me expecting all uCars to be Minecarts, will try to fix that (that part was for storage carts dropping their content)

I will do some more fiddling and the 1.16 branch is probably appropriate. I don’t wanna break the master branch.

Sorry for basically wasting your time with two pull requests. I‘m gonna limit myself

Jakllp commented:
Sorry did not mean to do that

storm345 commented:
Yeah makes sense. The isACar(..) function is what should be relied on to determine if it's a car instead of other checks (like is it a minecart) in other places. If there is an issue with storage carts then that code change ought to happen there so all 'is a car' checks are correct. By storage cart you mean the vanilla minecart with chest?

The getBehindBlock(...) it may be more appropriate to normalize the vector (make unit vector) and add it to find the block behind, although this is more costly for performance. The ideal implementation would be to generate a direction/unit vector from the cart velocity a fast way (avoiding square root) or re-organize the code to be able to re-use the normalized unit vector calculated for making the car move (Which could itself be more optimally generated performance wise...).

For checking the road block, if passing to the method accepting Block as the input didn't work then the method accepting Block wasn't correctly implemented for your revised identification system and that ought to have been changed and tested (along with any other places which identify/match blocks/items with the config) - instead of adding another 'overload' taking string which breaks the abstraction and uses a different implementation entirely. Fundamentally passing to a method with signature isARoadBlock(Block) doesn't impose any assumption about the config block identification scheme to the caller and should be able to be made to work.

What specifically was wrong with having air count as 'road block'? It is so that a car in mid-air is still able to be determined as a car. An entity changing whether or not it 'is a car' when it's in the air does not seem like the right behaviour.

Jakllp commented:
I just remembered why I removed air as a roadblock:
It would make a Minecart stop right after falling of a cliff. This is similar to the reason the behindblock check was added:

When a Minecart was used on rails and would then encounter a drop (used in Minecraft rollercoasters) it would just stop once it would detect air underneath the cart which would stop it on the edge of the cliff just sitting there. This is why I implemented the checkBehindBlock function. It keeps the Minecart from becoming a UCar when falling of a "rail cliff".
I then encountered the problem that the Minecart would just fall straight down as it became a uCar once it saw air beneath it without the block behind it being a rail. This caused it to loose all non-vertical velocity. Removing Air as a roadblock solved this.
Funnily enough it now still becomes a ucar for a minimal amount of time causing it to derail and generally act up when on rails after being dropped of such a cliff. 

I've now

Jakllp commented:
Working on support for older configs and I am a bit confused.
On first glance isPlacableOn(String materialName, byte data) and isBlockEqualToConfigIds(List<String> rawIds, Block block) do the same thing but with different inputs (especially strange that you want the materialName as a String instead of just handing over the block)
What exactly is the reason for this segregation?

I made isBlockEqualToConfigIds compatible with old configs and just calling it instead of isPlacableOn seems to work fine

storm345 commented:
> Working on support for older configs and I am a bit confused.
> On first glance isPlacableOn(String materialName, byte data) and isBlockEqualToConfigIds(List rawIds, Block block) do the same thing but with different inputs (especially strange that you want the materialName as a String instead of just handing over the block)
> What exactly is the reason for this segregation?
> 
> I made isBlockEqualToConfigIds compatible with old configs and just calling it instead of isPlacableOn seems to work fine

I think they do indeed do the same thing and it would be better to pass Block instead of materialName+data and re-use the same method. A lot of the code in this plugin is rather old and could be much better written so I wouldn't be surprised.

Jakllp commented:
Just copying the latest commit description
I obviously cannot create a 1.13+ branch but I think that would be the safest way if this code seems fine for now.
But that's obviously up to you

What has been done:
- Should work for 1.13+
- Support for legacy configs
- Metadata for cars on rails replacing getBehindBlock
- Teleport-Signs now support relative coordinates and are functional again
- Moved check for special minecarts to isACar -> Other types of uCars possible
- Updated ReadMe

Not tackled yet:
- Cannot enable custom hitboxes through NMS as width field of entities has seemingly been removed/changed

storm345 commented on src/main/java/com/useful/ucars/uCarsListener.java:
```diff
@@ -1540,11 +1592,11 @@ void interact(PlayerInteractEvent event) {
 				return;
 			}
-			if (!PlaceManager.placeableOn(iar.name().toUpperCase(), block.getData())) {
+			if (!plugin.isBlockEqualToConfigIds(roadBlocks, block)) { //replaces placableOn()
```
I think it would have been better here to call PlaceManager.placeableOn(Block) rather than breaking the abstraction and checking against the config

storm345 commented on src/main/java/com/useful/ucars/ucars.java:
```diff
@@ -683,21 +683,22 @@ public final Boolean isBlockEqualToConfigIds(List<String> rawIds, Block block) {
 			if (parts.length < 1) {
-			} else if (parts.length < 2) {
-				if (parts[0].equalsIgnoreCase(block.getType().name())) {
+			} else if (parts.length < 2) { //New configs and blocknames
```
This doesn't quite seem like the right check since the old format may not have specified the :data part but it's probably fine - just means some people will need to edit the configs a little

Jakllp replied:
I suspect you mean something like WOOD in the config and it therefore working on all types of wood?

storm345 replied:
Yeah that's what I mean

storm345 reviewed (commented):
Looks better now, thanks. If you can try using Entity.getWidth() (https://hub.spigotmc.org/javadocs/spigot/org/bukkit/entity/Entity.html#getWidth()) and see if that produces the correct value where it was previously doing reflection that would be great. Then it hopefully will work properly!

The 'was on rails' looks like an alright way to fix it and I think is probably better than what was there before but have you checked to make it sure it works properly when road blocks are not enabled? Most servers don't use the road blocks feature (and prefer to manage what is a uCar through uCarsTrade or make all minecarts uCars) and need to make sure there isn't an issue there.

If it works correctly when road blocks is disabled, and if entity.getWidth() works then I can make a 1.13plus branch and merge it into that! :)

Jakllp commented:
(if you can't really answer this question I will set up a 1.12 server and check):
Should the width for a Minecart return something like 0.98?

storm345 commented:
Sounds about right, if it drives okay it's probably fine. (If/whenever
uCarsTrade that sets the width is updated should find an issue then). If
you want to test it then install uCarsTrade on 1.12 with custom hitbox.

On Thu, 12 Nov 2020, 19:16 Jakl, <notifications@github.com> wrote:

> (if you can't really answer this question I will set up a 1.12 server and
> check):
> Should the width for a Minecart return something like 0.98?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/storm345dev/ucars/pull/9#issuecomment-726284058>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAU7ODJSS2HRDYDMMXJ6PFLSPQYCBANCNFSM4TOS2PWQ>
> .
>

Jakllp commented:
Should address all the things
Commit notes:

- Tested (and fixed) unenabled roadblocks
- Readded custom hitboxes (probably)
- placeableOn returns, better then ever
- Configs can now contain general blocktypes like "LOG" (ONLY blocks, not fuel items)

Fixes:
- Fixed a collision bug that caused errors and lags
- Fixed Maven-Build

storm345 commented on src/main/java/com/useful/ucars/PlaceManager.java:
```diff
@@ -29,5 +31,30 @@ public static Boolean placeableOn(String materialName, byte data) {
+			final String[] parts = raw.split(":");
+			if (parts.length < 1) {
+			} else if (parts.length < 2) { //New configs and blocknames
+				if (ItemStackFromId.equals(raw,block.getType().name().toUpperCase(),block.getData()) || block.getType().name().toUpperCase().contains(raw)) {
```
I'm not sure the contains(...) is the right check here. For example if you specify COBBLESTONE to be a road block you will also be inadvertently specifying COBBLESTONE_WALL and COBBLESTONE_STAIRS which you may not want

storm345 replied:
Why also can't this call to the "isBlockEqualToConfigIds" method to avoid code duplication?

Jakllp replied:
Mhh... To keep the required config changes as minimal as possible one could require something like "!WOOL" to add all wool varieties (or cobblestone varieties etc). Or do you have a better idea?

storm345 replied:
I think just leave it so that they have to specify all the ones the want? I don't see the problem with that

storm345 reviewed (changes requested):
Almost there!

storm345 commented on README.md:
```diff
@@ -1,6 +1,9 @@
-The repo of ucars(bukkit)
+1.13+ repo of uCars(bukkit)
+
+MISSING FEATURES:
```
Need to update the README :P

storm345 reviewed (commented):
Just the README needs updating, and then it's good to merge!

storm345 reviewed (approved):
I updated the readme, will merge it into branch '1.13plus' :)
